library/modules/EventManager.cpp

```diff
@@ -434,38 +539,50 @@ static void manageJobInitiatedEvent(color_ostream& out) {
     lastJobId = *df::global::job_next_id - 1;
 }
 
-static void manageJobStartedEvent(color_ostream& out){
+static void manageJobStartedEvent(color_ostream& out) {
     if (!df::global::world)
         return;
-
-    multimap<Plugin*,EventHandler> copy(handlers[EventType::JOB_STARTED].begin(), handlers[EventType::JOB_STARTED].end());
     int32_t tick = df::global::world->frame_counter;
+    int32_t oldest_last_tick = (uint16_t)-1;
 
-    // compile a list of newly started jobs
-    std::vector<df::job*> newly_started_jobs;
-    for ( df::job_list_link* link = df::global::world->jobs.list.next; link != NULL; link = link->next ) {
+    // update the started jobs list for the current tick
+    for (df::job_list_link* link = df::global::world->jobs.list.next; link != NULL; link = link->next) {
         df::job* job = link->item;
         // the jobs must have a worker to start
         if (job && Job::getWorker(job)) {
-            // cross-reference against existing list of jobs with workers
-            if (startedJobs.emplace(job).second) {
-                // must be new
-                newly_started_jobs.push_back(job);
+            // the job won't be added if it already exists
+            if (startedJobIDs.emplace(job->id).second) {
+                // DF is going to delete completed jobs, and we might not send these before they're completed
+                startedJobs.emplace(tick, Job::cloneJobStruct(job, true));
             }
         }
     }
     // iterate the event handlers
-    for(auto &iter : copy) {
+    multimap<Plugin*, EventHandler> copy(handlers[EventType::JOB_STARTED].begin(), handlers[EventType::JOB_STARTED].end());
+    for (auto &iter: copy) {
         auto &handler = iter.second;
+        auto last_tick = eventLastTick[handler.eventHandler];
+        oldest_last_tick = oldest_last_tick < last_tick ? oldest_last_tick : last_tick;
         // make sure the frequency of this handler is obeyed
-        if (tick - eventLastTick[handler.eventHandler] >= handler.freq) {
+        if (tick - last_tick >= handler.freq) {
             eventLastTick[handler.eventHandler] = tick;
-            // send the handler the new jobs
-            for(auto job : newly_started_jobs){
-                handler.eventHandler(out, (void*)job);
+            // send the handler all the new jobs since it last fired
+            auto jter = startedJobs.upper_bound(last_tick);
```

Comment:
Late listener registrations can result in a flood of old events from before the listener was registered. I'd comment in the register function, but it isn't in the diff.

Reply:
This could be desired behavior. Maybe we should have a flag on the register function to select whether the backlog is flushed or not

Reply:
sure, but I could add another param with a default value of off so nothing breaks.

Reply:
`bool backlog = false` was added to the signature in the header